### PR TITLE
Implementation of __init_subclass__

### LINKED ISSUE #1346: PEP 487 (__init_subclass__, __set_name__)
## Feature

`__init_subclass__`, `__set_name__`

## Python Documentation

https://docs.python.org/3/reference/datamodel.html#object.__init_subclass__
https://docs.python.org/3/reference/datamodel.html#object.__set_name__

### PR DESCRIPTION
**Pull request by BolunThompson**
This implements `__init_subclass__` from [PEP 487](https://www.python.org/dev/peps/pep-0487/).

Closes #1346. Partially fixes issue  #356.

### CODE REVIEW

File: vm/src/obj/objtype.rs [hunk 1]
```diff
@@ -327,6 +334,17 @@ impl PyClassRef {
...
 
+        if let Some(initter) = typ.get_super_attr("__init_subclass__") {
+            let initter = PyClassMethod::descr_get(
+                vm,
+                initter,
+                None,
+                OptionalArg::Present(typ.clone().into_object()),
+            )?;
+            let init_args = PyFuncArgs::from(kwargs);
+            vm.invoke(&initter, init_args)?;
```
Comment (coolreader18) [inline vm/src/obj/objtype.rs:345]: The second argument to `vm.invoke` can be any type that's `Into<PyFuncArgs>`:
```suggestion
            vm.invoke(&initter, kwargs)?;
```

File: vm/src/obj/objtype.rs [hunk 2]
```diff
@@ -327,6 +334,17 @@ impl PyClassRef {
             }
         }
 
+        if let Some(initter) = typ.get_super_attr("__init_subclass__") {
+            let initter = PyClassMethod::descr_get(
+                vm,
+                initter,
+                None,
+                OptionalArg::Present(typ.clone().into_object()),
+            )?;
```
Comment (coolreader18) [inline vm/src/obj/objtype.rs:343]: Maybe use `vm.call_descr_get_specific()` here? `initter` won't necessarily always be a PyClassMethod, even though that's not correct, but we should probably still call the appropriate `__get__` descriptor.

File: vm/src/obj/objtype.rs [hunk 3]
```diff
@@ -335,14 +335,10 @@ impl PyClassRef {
...
-                vm,
-                initter,
-                None,
-                OptionalArg::Present(typ.clone().into_object()),
-            )?;
-            let init_args = PyFuncArgs::from(kwargs);
-            vm.invoke(&initter, init_args)?;
+            let initter = vm
+                .call_get_descriptor_specific(initter, None, Some(typ.clone().into_object()))
+                .unwrap()?;
```
Comment (coolreader18) [inline vm/src/obj/objtype.rs:340]:
```suggestion
                .call_get_descriptor_specific(initter.clone(), None, Some(typ.clone().into_object()))
                .unwrap_or(Ok(initter))?;
```

**Comment by coolreader18** [comment]
Oh, actually, would you be able to put the copied CPython test file in a separate commit? Something like

- Add test_subclassinit.py from CPython 3.x.x
  - Same exact file as in CPython
- Mark expected failures
  - edit the test file to put `expectedFailure` et al

**Comment by BolunThompson** [comment]
I'm done with this.

**Comment by coolreader18** [comment]
Thank you for contributing!

**Pull request merged**